src/herder/HerderSCPDriver.cpp

```diff
@@ -351,7 +354,9 @@ HerderSCPDriver::validateValueHelper(uint64_t slotIndex, StellarValue const& b,
 
         res = SCPDriver::kInvalidValue;
     }
-    else if (!txSet->checkValid(mApp, 0))
+    else if (!txSet->checkValid(mApp, curProtocolContainsClosetimeChange()
```

Comment:
This seems suspicious to me. Suppose you have a transaction set containing a single transaction which is valid after time `T`. For a value with close-time `T+1`, this transaction is valid. But if the last closed ledger has close-time `T-1`, I think this check will erroneously consider this transaction invalid.

This is closely related to the "validity range" comments from @MonsieurNicolas.

Is there a test case for this? If so, which one covers it? If not, can you add one?

Reply:
By "valid after time `T`", you mean it has a `minTime` of `T`, correct?  If the last ledger has closetime `T-1`, that will indeed be considered invalid, specifically `txTOO_EARLY`.  And yes, I think I gather that that's what @MonsieurNicolas is objecting to as well.

What I don't understand is why that's erroneous.  That's exactly what we would have done before my change, isn't it?  And CAP-0034 only mentions `txTOO_LATE` and `maxTime`; it has zero mention of either `txTOO_LATE` or `minTime`.  And I don't understand why it _would_ mention `txTOO_EARLY` or `minTime`, because they do not suffer from the problems that `txTOO_LATE` and `maxTime` do which CAP-0034 is trying to address.

I also don't understand how a transaction set could even _reach_ this code if it contained a transaction with `minTime==T` in the presence of the condition `lcl==T-1` , because such a transaction would have been trimmed way back in `TransactionQueue::canAdd()`.  At that point, we don't _have_ any next-`closeTime` estimate, so the _only_ thing we can validate against is `lcl.closeTime`.  I therefore don't understand how adding a bunch of code to allow transactions with `minTime > lcl.closeTime` at this point would even change the system's semantics.

If there is actually some reason that it would make a difference to validate `minTime`s against the `closeTime` in the `StellarValue` at this point, then I not only have to write a bunch more code, but also a bunch more documentation, because I did not understand that the CAP that I myself was writing implied a change to the validation of `minTime` as well as that of `maxTime`.  (There's no occurrence of either `minTime` or `txTOO_EARLY` in the discussion in the entire discussion in the [issue](https://github.com/stellar/stellar-protocol/issues/622), either.)

Reply:
It is not exactly what we would have done before. Here is an example. Suppose you have a transaction set containing a single transaction which is valid after time `T` and before time `T` (I believe both the time bounds are inclusive, but if that's not true then just assume it for the sake of this example). For a value with close-time `T`, this transaction should be valid. But with your approach it is _impossible_ for this transaction set to be valid, because the close-time of two consecutive ledgers must differ by at least one second. In the existing protocol, this transaction set is at least valid if the previous close-time was exactly `T`.

Furthermore, as @MonsieurNicolas noted, you are conflating implementation details with protocol details. What stops me from modifying stellar-core to accept transactions that I anticipate will be valid at the next ledger close? Nothing, because that is an implementation detail. But once I have done that, I could easily include such transactions in a transaction set that I nominate. Because I am nominating a <close-time, txset> pair, such a transaction set should be valid if every transaction in the set would be valid at that close-time.

The key point here is that what we were doing before is wrong. That's the whole reason that CAP-34 exists. Just because your proposed change makes the situation better than it was before, doesn't mean it's the best possible change. It is the duty of the author to evaluate all of the potential consequences of the proposal, and thereby achieve the best properties that are reasonable. From my CAP-15 blog post https://www.stellar.org/developers-blog/how-we-simplified-fees-on-the-stellar-network:

> First, even changes that sound simple, such as making it possible to increase the fee on a transaction, can be extremely challenging to design and implement. Second, we need to strike the right balance between providing the strongest guarantees that we can and designing features which can actually be implemented safely.

In this case, I think providing stronger guarantees will make the implementation easier to reason about and therefore safer.

Reply:
> It is not exactly what we would have done before. Here is an example. Suppose you have a transaction set containing a single transaction which is valid after time `T` and before time `T` (I believe both the time bounds are inclusive, but if that's not true then just assume it for the sake of this example). For a value with close-time `T`, this transaction should be valid. But with your approach it is _impossible_ for this transaction set to be valid, because the close-time of two consecutive ledgers must differ by at least one second. In the existing protocol, this transaction set is at least valid if the previous close-time was exactly `T`.

I think I understand you to be saying that the transaction has `minTime==T` and `maxTime==T`, and the `StellarValue` `v` that the transaction is in has `v.closeTime==T`.  Correct?

But the old code here validated `minTime` and `maxTime` against `lcl.closeTime`, and with `minTime==maxTime`, `lcl.closeTime==minTime==maxTime` is the only way to avoid both `txTOO_EARLY` and `txTOO_LATE`.  Correct?

And as you point out, for a `StellarValue` `v` to be valid, it must also have `v.closeTime > lcl.closeTime` (strict inequality).  But we have also shown that validity requires `lcl.closeTime==minTime==T==v.closeTime`.  This is a contradiction.  So the old code would also have failed the validity check here, specifically when it tried to validate that `v.closeTime` was strictly greater than `lcl.closeTime`.  Correct?

> Furthermore, as @MonsieurNicolas noted, you are conflating implementation details with protocol details. What stops me from modifying stellar-core to accept transactions that I anticipate will be valid at the next ledger close? Nothing, because that is an implementation detail. But once I have done that, I could easily include such transactions in a transaction set that I nominate. Because I am nominating a <close-time, txset> pair, such a transaction set should be valid if every transaction in the set would be valid at that close-time.

The protocol also doesn't require that we _do_ accept transactions that weren't valid at the last ledger close, does it?  In particular, the existing code in _our_ stellar-core would have rejected such a transaction, so it seems to me that that surely wouldn't be a violation of the protocol.  In other words, for some other implementation to accept transactions that it _anticipates_ to be valid (i.e. not too early) at the time of the next ledger close is that implementation's free choice of implementation detail, and for our implementation _not_ to accept such a transaction unless it was already valid (i.e. not too early) at the time of last ledger close is _our_ free choice of implementation detail.

> The key point here is that what we were doing before is wrong. That's the whole reason that CAP-34 exists. Just because your proposed change makes the situation better than it was before, doesn't mean it's the best possible change. It is the duty of the author to evaluate all of the potential consequences of the proposal, and thereby achieve the best properties that are reasonable. From my CAP-15 blog post https://www.stellar.org/developers-blog/how-we-simplified-fees-on-the-stellar-network:
> 
> > First, even changes that sound simple, such as making it possible to increase the fee on a transaction, can be extremely challenging to design and implement. Second, we need to strike the right balance between providing the strongest guarantees that we can and designing features which can actually be implemented safely.
> 
> In this case, I think providing stronger guarantees will make the implementation easier to reason about and therefore safer.

I agree -- and I'm arguing for a stronger guarantee.  Changing my code to validate `minTime` against `v.closeTime` instead of `lcl.closeTime`, as I gather you and Nicolas are suggesting, would make the guarantees that we make about valid `StellarValue`s strictly weaker (wouldn't it?).

Reply:
See [this comment](https://github.com/stellar/stellar-core/pull/2625#discussion_r458379008) for the upcoming resolution.